### ISSUE TITLE
DataSource metrics are not auto-configured when using lazy initialization

### ISSUE THREAD

**Issue by bellatoris** [spring-projects/spring-boot#30282]
spring boot version: 2.6.4

After setting `spring.main.lazy-initialization=true`, hikari related metric disappeared from metric list. Although HikariDataSource is created during application context initialization, it was confirmed that HikariDataSourceMetricsConfiguration is not set. I have confirmed that HikariDataSourceMetricsConfiguration is the target of autoconfiguration as follows.
```
   DataSourcePoolMetadataProvidersConfiguration.HikariPoolDataSourceMetadataProviderConfiguration matched:
      - @ConditionalOnClass found required class 'com.zaxxer.hikari.HikariDataSource' (OnClassCondition)

   DataSourcePoolMetricsAutoConfiguration matched:
      - @ConditionalOnClass found required classes 'javax.sql.DataSource', 'io.micrometer.core.instrument.MeterRegistry' (OnClassCondition)
      - @ConditionalOnBean (types: javax.sql.DataSource,io.micrometer.core.instrument.MeterRegistry; SearchStrategy: all) found beans 'prometheusMeterRegistry', 'dataSource' (OnBeanCondition)
```

**Comment by wilkinsona** [comment]
DataSource metrics are currently bound via some slightly unusual `@Autowired` methods and it's been that way since the code [was introduced](https://github.com/spring-projects/spring-boot/commit/9b6f0c83bf684dcb2b78c4247cde0436a2495c00).

If `MeterBinder` implementations were used instead, I think this would work correctly with lazy initialization. Perhaps there's a reason why a `MeterBinder` couldn't be used? Each`MeterBinder` is called as part of post-processing of the `MeterRegistry` bean which is earlier than the `@Autowired` approach while will happen sometime after the registry has been fully initialized. Flagging for team attention to see if anyone can recall if there was a reason for the current approach.

**Comment by snicoll** [comment]
In the current arrangement we have a `@Configuration` class with no hook-point to do something so we chose to add `@Autowired` for the lack of a better option. I agree it is unusual. 

We considered switching those to implement `SmartInitializingSingleton`at some point but I can't find back the issue where this was discussed (if any). IMO, anything that gets rid of those and better use Micrometer's infrastructure is an improvement.

**Comment by wilkinsona** [comment]
Thanks, @snicoll. A `MeterBinder`-based approach seems to work fine. [These changes](https://github.com/spring-projects/spring-boot/issues/30282) build cleanly at least. I am quite tempted to merge them.

**Comment by philwebb** [comment]
We're going to merge the above fix, but only in 2.7 since there's some risk involved. For 2.5 and 2.6, please use [the workaround above](https://github.com/spring-projects/spring-boot/issues/30282#issuecomment-1072449335)
